Add subdivisions in ticksscalebar right segments

agiudiceandrea opened the pull request:
## Description

Adds the ability to create subdivisions for segments included in the right part of the ticksscalebar.

Line Ticks Up
![image](https://user-images.githubusercontent.com/16253859/80860629-47116b00-8c69-11ea-83d8-61f27cb77e68.png)

Line Ticks Middle
![image](https://user-images.githubusercontent.com/16253859/80860663-6c9e7480-8c69-11ea-94a3-8d0e8544867d.png)

Line Ticks Down
![image](https://user-images.githubusercontent.com/16253859/80860675-7b852700-8c69-11ea-984d-584eff1e3c95.png)

Scalebar item properties - Segments groupbox
![image](https://user-images.githubusercontent.com/16253859/81009703-653bbe80-8e55-11ea-9286-4850b32edb23.png)




I have not changed the behaviour related to the segments on the left side of the scalebar.

<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

nyalldawson commented on src/core/layout/qgslayoutitemscalebar.h:
```diff
@@ -81,6 +81,18 @@ class CORE_EXPORT QgsLayoutItemScaleBar: public QgsLayoutItem
+    /**
+     * Returns the number of subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see setNumberOfSubdivisions()
+     */
```
\since QGIS 3.14

nyalldawson commented on src/core/layout/qgslayoutitemscalebar.h:
```diff
@@ -81,6 +81,18 @@ class CORE_EXPORT QgsLayoutItemScaleBar: public QgsLayoutItem
+    /**
+     * Sets the number of \a subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see numberOfSubdivisions()
+     */
```
\since QGIS 3.14

nyalldawson commented on src/core/scalebar/qgsscalebarsettings.h:
```diff
@@ -120,6 +120,18 @@ class CORE_EXPORT QgsScaleBarSettings
+    /**
+     * Returns the number of subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see setNumberOfSubdivisions()
+     */
```
\since QGIS 3.14

nyalldawson commented on src/core/scalebar/qgsscalebarsettings.h:
```diff
@@ -120,6 +120,18 @@ class CORE_EXPORT QgsScaleBarSettings
+    /**
+     * Sets the number of \a subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see numberOfSubdivisions()
+     */
```
\since QGIS 3.14

nyalldawson commented on src/core/scalebar/qgsticksscalebarrenderer.cpp:
```diff
@@ -114,13 +115,44 @@ void QgsTicksScaleBarRenderer::draw( QgsRenderContext &context, const QgsScaleBa
+        break;
+      case TicksMiddle:
+        verticalPos = middlePosition;
+        subTickPositionsY << (barTopPosition + middlePosition ) / 2;
```
I think ideally we'd expose the subdivision height to user control

agiudiceandrea replied:
Hi @nyalldawson, thanks for the revision of this preliminary work.
What do you think about letting the user choose the subdivisions tick height <= the segments tick height?

nyalldawson replied:
I think it's a good idea. We could either expose it as an absolute mm size or a percentage (of tick height) size...

agiudiceandrea replied:
Subdivisions tick height control added.

nyalldawson reviewed (commented):
Code is looking good! Got some screenshots to share?

agiudiceandrea committed:
[feature] Add subdivisions in ticksscalebar right segments

Adds the ability to create subdivisions for segments included in the right part of the ticksscalebar

agiudiceandrea committed:
Expose the subdivisions tick height to user control in ticksscalebar

nyalldawson commented:
Nice! Looking good to me. Can you add a unit test and then we can merge....

agiudiceandrea committed:
Add test for subdivisions in ticksscalebar right segments
